### PR TITLE
QCBM algorithm

### LINKED ISSUE #264: Adding new examples and algorithms for TQ
To enrich torchquantum's support of various algorithms, we need your help on implementing more quantum algorithms and examples in TQ! You may choose any quantum algorithm you like to implement such as QSVT or generative quantum models. Once complete, add a Python file and notebook to `examples` and (if you see fit) an algorithm to `torchquantum/algorithms`!

Please don't hesitate to let us know for any questions!

### PR DESCRIPTION
**Pull request by Gopal-Dahale**
## Purpose
Added Quantum Circuit Born Machine (QCBM) algorithm.

## Details
- QCBM and MMDLoss classes are added in `torchquantum/algorithms`.
- Added an example demonstrating generative modelling with QCBM on Gaussian Mixture.
- Resolves #264.

## Mentions
@GenericP3rson

### CODE REVIEW

**Comment by Gopal-Dahale** [comment]
@GenericP3rson Can you help me with this `flake8` error in github actions.

**Comment by GenericP3rson** [comment]
Hi, don’t worry about getting that working, you just need to pass the functional tests!

File: examples/QCBM/README.md [hunk 1]
```diff
@@ -0,0 +1,8 @@
+# Quantum Circuit Born Machine
+
```
Comment (GenericP3rson) [inline examples/QCBM/README.md:2]: You should definitely add attribution (Implementation By: [Your Name])!

File: examples/QCBM/qcbm_gaussian_mixture.py
Comment (GenericP3rson) [inline examples/QCBM/qcbm_gaussian_mixture.py:1]: Two format suggestions:
(1) Can you pass command line parameters to this example? (Specifically, use `argparse` to add parameters on whether or not to plot, n_wires, epochs, and essentially all of the mixture parameters — you can find an example in `examples/mnist/mnist.py`.) You’ll also need to add a simple description on how to use it in the README. The idea of this is we want to be able to run some MVP example in our CI to make sure the example stays up to date

(2) Also, if you could make this example into a simple notebook (no need to add parameters for the notebook itself), that would be great! Adding a bit of text to explain what’s going on in the notebook would be useful as well!

**Review by GenericP3rson — CHANGES_REQUESTED** [review]
Great work so far! Main thing to change is formatting, see comments below. Thanks for the great work and feel free to reach out with questions!

File: examples/QCBM/qcbm_gaussian_mixture.py
Comment (Gopal-Dahale) [reply examples/QCBM/qcbm_gaussian_mixture.py:1]: @GenericP3rson 
1. Added command line arguments to the Python file and updated the readme.
2. I think adding a jupyter notebook will result in code duplication. If it works, I can add some text in the Python file. I have added a visualization of the result (plot of learnt Gaussian mixture) in the README.md. Let me know your on this part.

Comment (GenericP3rson) [reply examples/QCBM/qcbm_gaussian_mixture.py:1]: (By the way, thanks for your patience! You raise a really excellent point on code duplication; I’m going to touch base with the rest of the team to decide a standard for our examples tonight then will get back to you!)

Comment (GenericP3rson) [reply examples/QCBM/qcbm_gaussian_mixture.py:1]: Hello, I discussed with the team! We think having a notebook to introduce the background and functionality would be very helpful to users. To help reduce code replication, you can try importing some of the functions from the Python file. And if you must replicate code, that’s also totally fine by us!

By the way, really, really great work so far!

Comment (Gopal-Dahale) [reply examples/QCBM/qcbm_gaussian_mixture.py:1]: @GenericP3rson I have added a jupyter notebook which introduces QCBM.

File: torchquantum/algorithm/qcbm.py [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+        # Kernel matrix
+        self.K = sum(torch.exp(-gamma * sq_dists) for gamma in gammas) / len(scales)
+        self.scales = scales
+
+    def k_expval(self, px, py):
+        """
+        Kernel expectation value
+
+        Args:
+			px: First probability distribution
```
Comment (GenericP3rson) [inline torchquantum/algorithm/qcbm.py:36]:
```suggestion
		px: First probability distribution
```

File: torchquantum/algorithm/qcbm.py [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+			py: Second probability distribution
```
Comment (GenericP3rson) [inline torchquantum/algorithm/qcbm.py:37]:
```suggestion
		py: Second probability distribution
```

File: torchquantum/algorithm/qcbm.py [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+        Returns:
+            Expectation value of the RBF Kernel.
+        """
+
+        return px @ self.K @ py
+
+    def forward(self, px, py):
+        """
+        Squared MMD loss.
+
```
Comment (GenericP3rson) [inline torchquantum/algorithm/qcbm.py:48]:
```suggestion
    Args:
```

File: torchquantum/algorithm/qcbm.py [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+        px: First probability distribution
```
Comment (GenericP3rson) [inline torchquantum/algorithm/qcbm.py:49]:
```suggestion
            px: First probability distribution
```

File: torchquantum/algorithm/qcbm.py [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+        py: Second probability distribution
```
Comment (GenericP3rson) [inline torchquantum/algorithm/qcbm.py:50]:
```suggestion
            py: Second probability distribution
```

**Review by GenericP3rson — CHANGES_REQUESTED** [review]
Great work! A few last minor changes. If you could just keep the tabbing consistent in the qcbm.py file, that would be great! (A few examples are suggested—note that there might be some oddities with the spaces converting properly into tabs. I think it would be most optimal to just double-check that everything is consistent in the file itself.)

**Comment by Gopal-Dahale** [comment]
@GenericP3rson what do you think about formatting now?

**Review by GenericP3rson — APPROVED** [review]
Good to go!

**Review by Hanrui-Wang — APPROVED** [review]

**Pull request merged**